core/src/wsv.rs

```diff
@@ -1044,6 +1047,140 @@ impl WorldStateView {
+
+        let world = self.world_mut();
+        if world
+            .permission_token_definitions
+            .contains_key(&definition_id)
+        {
+            return Err(RepetitionError {
+                instruction_type: InstructionType::Register,
+                id: IdBox::PermissionTokenDefinitionId(definition_id),
+            }
+            .into());
+        }
+
+        world
+            .permission_token_definitions
+            .insert(definition_id, definition.clone());
+
+        self.emit_events(Some(PermissionTokenEvent::DefinitionCreated(definition)));
+
+        Ok(())
+    }
+
+    /// Unregister [`PermissionTokenDefinition`].
+    ///
+    /// # Errors
+    ///
+    /// Fails with [`FindError`] if [`PermissionTokenDefinition`] with provided
+    /// `definition_id` does not exist.
+    pub fn unregister_permission_token_definition(
+        &mut self,
+        definition_id: PermissionTokenId,
+    ) -> Result<(), Error> {
+        self.remove_token_from_roles(&definition_id)?;
+        self.remove_token_from_accounts(&definition_id)?;
+
+        let world = self.world_mut();
+        let definition = world
+            .permission_token_definitions
+            .remove(&definition_id)
+            .ok_or_else(|| FindError::PermissionTokenDefinition(definition_id))?;
+
+        self.emit_events(Some(PermissionTokenEvent::DefinitionDeleted(definition)));
+
+        Ok(())
+    }
+
+    /// Remove all tokens with specified definition id from all registered roles
+    fn remove_token_from_roles(
+        &mut self,
+        target_definition_id: &<PermissionTokenDefinition as Identifiable>::Id,
+    ) -> Result<(), Error> {
+        let mut roles_containing_token = Vec::new();
+
+        for (role_id, role) in self.roles().iter() {
+            if role
+                .permissions
+                .iter()
+                .any(|token| token.definition_id == *target_definition_id)
+            {
+                roles_containing_token.push(role_id.clone())
+            }
+        }
+
+        let mut events = Vec::with_capacity(roles_containing_token.len());
+        let world = self.world_mut();
+        for role_id in roles_containing_token {
+            if let Some(role) = world.roles.get_mut(&role_id) {
+                role.permissions
+                    .retain(|token| token.definition_id != *target_definition_id);
+                events.push(RoleEvent::PermissionRemoved(PermissionRemoved {
+                    role_id: role_id.clone(),
+                    permission_definition_id: target_definition_id.clone(),
+                }));
+            } else {
+                error!(%role_id, "role not found. This is a bug");
+                return Err(FindError::Role(role_id.clone()).into());
+            }
+        }
+
+        self.emit_events(events);
+
+        Ok(())
+    }
+
+    /// Remove all tokens with specified definition id from all accounts in all domains
```

Comment:
I'm a bit concerned with leaving this in long-term. 

Hopefully we'll have some opportunity to discuss the new Iroha ISA, and these functions will be generic.

Reply:
What does *ISA* stand for?